README.md
# NanoScatterFit


[![Pypi link](https://img.shields.io/pypi/v/nanoscatterfit.svg)](https://pypi.python.org/pypi/nanoscatterfit)
[![Travis job](https://img.shields.io/travis/plunkettal/nanoscatterfit.svg)](https://travis-ci.org/plunkettal/nanoscatterfit)




A simple package for analyzing nanoparticle dispersions and superstructures

## Table of Content:

- [Intallation](#installation)
- [Usage](#usage)
- [TODO](#todo)
- [Contributing](#contributing)
- [Credits](#credits)

## Installation

**Note:** below pip installation is just a placeholder for first stable release on PyPI. For now download the latest version from the GitHub repository. 
```batch

    $ pip install nanoscatterfit
```

This is the preferred method to install NanoScatterFit, as it will always
install the most recent stable release.

If you don't have [pip](https://pip.pypa.io) installed, this 
[Python installation guide](http://docs.python-guide.org/en/latest/starting/installation/) 
can guide you through the process.

## Usage

- import required functions from the module nanoscatterfit. 
- adjust functions as required for own datasets
- examples can be found [here](./examples/)


## TODO

- [ ] Upload first stable version on Pypi 
- [ ] Update Unit tests


## Contributing

Contributions are welcome, and they are greatly appreciated! Every
little bit helps, and credit will always be given.

For more info please click [here](./CONTRIBUTING.md)


## Credits

This package was created with Cookiecutter and the `oldani/cookiecutter-simple-pypackage` project template.

- [Cookiecutter](https://github.com/audreyr/cookiecutter)
- [oldani/cookiecutter-simple-pypackage](https://github.com/oldani/cookiecutter-simple-pypackage)


HISTORY.md
# History
# 0.1.2 (2023-12-07)

- added kwarg min_q111 for function fit_structurefactor()
- added return of r_squared in fit_formfactor()
- log filename changed to 'nanoscatterfit.log'
- delete_log_file() doesnt remove the file anymore but just erases its content in order to prevent it from not being recreated in a running jupyter notebook file
- function isscatter() now also checks if the 111 peak intensity is unexpectedly small


# 0.1.1 (2023-11-25)

- fixed issues with detecting irrelevant peaks as the (111) peak in the function plot_structuremodel()
- added examples for batch analyzing suspensions and superstructures

# 0.1.0 (2023-11-15)

- First release on GitHub.




tests/test_nanoscatterfit.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#%%
"""Tests for `nanoscatterfit` package."""

import pytest
import pandas as pd
import os

from nanoscatterfit  import *


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    
    # assert 'GitHub' in BeautifulSoup(response.content).title.string
    
def test_import_rawdata():
    # Create a sample .txt file
    sample_data = "q I\n0.1 10\n0.2 20\n0.3 30"
    sample_file = 'test_data.txt'
    with open(sample_file, 'w') as file:
        file.write(sample_data)

    # Call the function with the sample file
    result_df = import_rawdata(sample_file)

    # Check if the result is as expected
    expected_df = pd.DataFrame({'q': [0.1, 0.2, 0.3], 'I': [10, 20, 30]})
    pd.testing.assert_frame_equal(result_df, expected_df)

    # Clean up: remove the sample file after test
    os.remove(sample_file)


def test_auto_baseline_optimizer_empty_input():
    y = pd.Series([])
    baseline = auto_baseline_optimizer(y)
    assert baseline is None


def test_auto_baseline_optimizer_single_value_input():
    y = pd.Series([1])
    baseline = auto_baseline_optimizer(y)
    assert baseline is None


def test_auto_baseline_optimizer_monotonic_input():
    y = pd.Series([1, 2, 3, 2, 1])
    baseline = auto_baseline_optimizer(y)
    assert np.allclose(baseline, [1, 1, 1, 2, 2])


def test_auto_baseline_optimizer_non_monotonic_input():
    y = pd.Series([3, 2, 1, 4, 3])
    baseline = auto_baseline_optimizer(y)
    assert np.allclose(baseline, [3, 2, 1, 3, 3])


def test_cost_func_one_peak():
    y = pd.Series([1, 2, 3, 2, 1])
    baseline = np.array([1, 1, 1, 2, 2])
    cost = cost_func(y, baseline)
    assert cost == 1


def test_cost_func_two_peaks():
    y = pd.Series([3, 2, 1, 4, 3])
    baseline = np.array([3, 2, 1, 3, 3])
    cost = cost_func(y, baseline)
    assert cost == 2


def test_filter_monotonic_peaks_one_peak():
    y = pd.Series([1, 2, 3, 2, 1])
    peaks = np.array([1, 3, 4, 5, 6])
    filtered_peaks = filter_monotonic_peaks(y, peaks)
    assert np.allclose(filtered_peaks, [1, 3, 4, 5])


def test_filter_monotonic_peaks_two_peaks():
    y = pd.Series([3, 2, 1, 4, 3])
    peaks = np.array([1, 3, 4, 5, 6])
    filtered_peaks = filter_monotonic_peaks(y, peaks)
    assert np.allclose(filtered_peaks, [1, 3, 4, 5])


def test_fit_structurefactor_fcc():
    x = pd.Series([0.1, 0.2, 0.3, 0.4, 0.5])
    y = pd.Series([1, 2, 3, 4, 5])
    result = fit_structurefactor(x, y, 'fcc')
    assert result.success
    assert result.params['cen1'].value == 0.3
    assert result.params['sigma1'].value == 0.03


def test_fit_structurefactor_unknown_structure():
    x = pd.Series([0.1, 0.2, 0.3, 0.4, 0.5])
    y = pd.Series([1, 2, 3, 4, 5])
    result = fit_structurefactor(x, y, 'unknown')
    assert not result.success


def test_fit_formfactor_lognormal_spherical():
    q_data = pd.Series([0.1, 0.2, 0.3, 0.4, 0.5])
    I_data = pd.Series([1, 2, 3, 4, 5])
    result = fit_formfactor(q_data, I_data, distribution='lognormal', shape='spherical')
    assert result.success
    assert result.params['mu'].value == 1.3
    assert result.params['sigma'].value == 0.1


# def test_fit_formfactor_normal_cubic():
#     q_data = pd.Series([0.1, 0.2, 0.3, 0.4, 0.5])
#     I_data = pd.Series([1, 2, 3, 4, 5])
#     result = fit_formfactor(q_data, I_data, distribution='
# %%


nanoscatterfit/utils.py
import pandas as pd
import os
from typing import Tuple, Optional,Dict,Any,List
from scipy import integrate
from scipy.signal import find_peaks
import logging
logging.basicConfig(
    level=logging.INFO,
    filename='nanoscatterfit.log',
    filemode='w',
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
from functools import wraps





#* general functions

def log_function_call(func):
    '''
    automatically logs used functions when @log_function_call is used above a function.
    '''
    @wraps(func)
    def wrapper(*args, **kwargs):
        # Log function call with trimmed arguments
        trimmed_args = [type(arg) if len(str(arg))>100 else arg for arg in args ]
        trimmed_kwargs = {k: type(v) if len(str(v))>100 else v for k, v in kwargs.items()}
        logging.info(f"Called function {func.__name__} with args: {trimmed_args} and kwargs: {trimmed_kwargs}")
        try:
            result = func(*args, **kwargs)
        except Exception as e:
            logging.exception(f"Function {func.__name__} raised an exception")
            raise
        # Log return value, trimmed if it's a DataFrame
        trimmed_result = type(result) if len(str(result))>100 else result
        logging.info(f"Function {func.__name__} returned {trimmed_result}")
        return result
    return wrapper


def delete_log_file(file_path: str='nanoscatterfit.log') -> None:
    """
    Deletes the specified log file.

    Parameters:
    - file_path : str
        The path to the log file to be deleted.

    Returns:
    - None
    """
    # logging.shutdown()
    try:
        
        with open(file_path, 'w') as f:
            #rewrite the content of the log file with the empty sting
            f.write('')
            f.close()
        print(f"Log file {file_path} has been deleted.")
    except FileNotFoundError:
        print(f"File {file_path} not found.")
    except PermissionError:
        print(f"Permission denied. Could not delete file {file_path}.")
    except Exception as e:
        print(f"An error occurred: {e}")
def check_columns(df: pd.DataFrame, required_columns: list, file_name: str):
    """Check if required columns exist in the DataFrame."""
    for col in required_columns:
        if col not in df.columns:
            raise ValueError(f"Column {col} is missing in the file {file_name}.")


#* data processing


@log_function_call
def cut_diffractogram(df: pd.DataFrame, qmin: float = 0, qmax: float = 1.5) -> Optional[pd.DataFrame]:
    """
    Cuts the diffractogram into the desired q range specified by qmin and qmax.

    Parameters:
    - df : pd.DataFrame
        DataFrame containing the diffractogram data. Must have a 'q' column.
    - qmin : float, optional
        Minimum q value for the diffractogram cut. Default is 0.
    - qmax : float, optional
        Maximum q value for the diffractogram cut. Default is 1.5.

    Returns:
    - pd.DataFrame or None
        DataFrame containing the cut diffractogram, or None if the input DataFrame is empty or invalid.
    """

    if df.empty or 'q' not in df.columns:
        logging.warning("Input DataFrame is empty or does not contain a 'q' column.")
        return None

    cut_df = df.loc[(df['q'] > qmin) & (df['q'] < qmax)].reset_index(drop=True)
    if cut_df.empty:
        logging.warning(f"No data points found in the range qmin={qmin}, qmax={qmax}.")
        return None

    return cut_df

# @log_function_call
def isscatter(df: pd.DataFrame, intensitytheshhold: int = 20) -> Optional[bool]:
    """
    Checks if the file is actually a scattering file by finding whether the first peak is in a reasonable q range.
    Adjust according to data structure.

    Parameters:
    - df : pd.DataFrame
        DataFrame containing the spectrum data. Must have a 'q' and 'I' column.
    - qmin : float, optional
        Minimum q value for checking the peak. Default is 0.
    - qmax : float, optional
        Maximum q value for checking the peak. Default is 0.3.

    Returns:
    - bool or None
        True if the first peak is within the specified q range, False otherwise.
    """

    if df.empty or 'q' not in df.columns or 'I' not in df.columns:
        logging.warning("Input DataFrame is empty or does not contain 'q' and/or 'I' columns.")
        return None

    # logging.info(f"Trying to check if the diffractogram has a peak between {qmin} and {qmax}")
    try:
        id_max = df['I'].idxmax()
        peak_q_value = df.loc[id_max, 'q']
        peaks=find_peaks(df.S, prominence=0.1)
        # if qmin < peak_q_value < qmax:
        logging.info('Checking if intensity at larger q values is at ca. 10')
        meanintensity=df.loc[(df.q > 2* peak_q_value)&(df.q < 5* peak_q_value)].I.mean()
        if meanintensity < intensitytheshhold:
            logging.info(f'The mean intensity of the diffractogram is {meanintensity:.0f}, and presumably not a diffraction file.')
            return False
        elif peaks[0][0] != df.S.idxmax():
            logging.info('Checking if 111 peak is largest')
            
            logging.info(f'The 111 peak was unexpectedly small.')
            return False
        else:
            logging.info(f'The mean intensity of the diffractogram is {meanintensity:.0f}, and presumably a diffraction file.')
            return True
    except Exception as e:
        logging.exception(f"An error occurred: {e}")
        return False
    # else:
    #     logging.info(f"The peak is not in the range between {qmin} and {qmax}. It's at q = {peak_q_value}.")
    #     return False
    

def normalize_area(x, y):
    """
    Normalizes the area under the curve represented by x and y.

    Parameters:
    - x : array-like
        The x values.
    - y : array-like
        The y values.

    Returns:
    - array-like
        The y values normalized such that the area under the curve is 1.
    """
    integral = integrate.trapz(y, x)
    return y / integral



@log_function_call
def auto_cutdiffr(x:pd.Series,y:pd.Series,maxpeakposition:float)->Optional[pd.Series]:
    '''
    cuts the diffractogram at the first and last minimum to improve fitting process.
    Ideally data is normalized to area=1 and smoothed.
    
    Parameters:
    - y : pd.Series
        Series containing the diffractogram data.

    Returns:
    - pd.Series 
        reduced diffractorgram
    '''
    
    y=y.loc[(x < maxpeakposition)]
    x=x.loc[(x < maxpeakposition)]
    logging.info(f'precut from {x.iloc[0]} to {x.iloc[-1]}')
    # logging.info(y.iloc[-5:])
    for i in (y.index):
        # if i==0:continue
        if (y[i] <= y[i+1])&(y[i] < y[i+2]):
            start=i
            break
    for i in reversed(y.index):

        if (y[i] <= y[i-1])&(y[i] < y[i-2]):
            y[i]
            end=i
            break
    logging.info(f'start={x[start]}, end={x[end]}')
    if start==end:
        logging.error(f'couldnt autocut diffractogram')
        return None,None
    return x[start:end],y[start:end]

nanoscatterfit/nanoscatterfit.py
# -*- coding: utf-8 -*-
"""Main module."""

#%%

import pandas as pd

from typing import Tuple, Optional, Union,Dict,Any,List
from scipy import interpolate, sparse
from scipy.sparse.linalg import spsolve
from scipy.optimize import curve_fit
from scipy.signal import find_peaks
from lmfit import Minimizer, Parameters
from lmfit.minimizer import MinimizerResult
from lmfit.lineshapes import gaussian, lognormal


import os
import numpy as np
import pandas as pd

import matplotlib
import matplotlib.pyplot as plt



from .utils import *
from .dataimport import *

    


@log_function_call
def calc_structurefactor(sampledata: pd.DataFrame, suspdata: pd.DataFrame, normalize:bool=False)-> pd.DataFrame:
    """
    Calculates the structure factor and adds it as column to the sampledata 

    Parameters:
    - sampledata : pd.DataFrame
        Sample data containing at least a 'q' and 'I' column.
    - suspdata : pd.DataFrame
        Suspension data containing at least a 'q' and 'I' column.
    - normalize : bool, optional
        normalizes area underneath the 'q' and 'S' columns to 1 if True. Default is False.
    Returns:
    - sampledata: pd.DataFrame
        dataframe of sample augmented by the form factor and structure factor column.
    """
    logging.info('Calculating structure factor')

    if 'q' not in sampledata.columns or 'I' not in sampledata.columns:
        logging.error("Sample data is missing required columns.")
        return

    if 'q' not in suspdata.columns or 'I' not in suspdata.columns:
        logging.error("Suspension data is missing required columns.")
        return

    f = interpolate.interp1d(suspdata['q'], suspdata['I'], kind='cubic')
    if (sampledata['q'].iloc[0] < suspdata['q'].iloc[0]) | (sampledata['q'].iloc[-1] < suspdata['q'].iloc[-1]):
        logging.warning("Sample data starts or ends before suspension data. sampledata will be cut.")
        sampledata= cut_diffractogram(sampledata,suspdata['q'].iloc[0],suspdata['q'].iloc[-1])
    sampledata['F'] = pd.Series(f(sampledata['q']))
    sampledata['S'] = sampledata['I'] / sampledata['F']
    if normalize==True:
        sampledata['S'] = normalize_area(sampledata['q'], sampledata['S'])
    return sampledata
@log_function_call
def baseline_als_optimized(y, lam, p, niter=50):
    ''''
    baseline ALS fit which works quite well for spectroscopic and diffraction data
    '''
    L = len(y)
    D = sparse.diags([1,-2,1],[0,-1,-2], shape=(L,L-2))
    D = lam * D.dot(D.transpose()) # Precompute this term since it does not depend on `w`
    w = np.ones(L)
    W = sparse.spdiags(w, 0, L, L)
    for i in range(niter):
        W.setdiag(w) # Do not create a new matrix, just update diagonal values
        Z = W + D
        z = spsolve(Z, w*y)
        w = p * (y > z) + (1-p) * (y < z)
    return z



@log_function_call    
def auto_baseline_optimizer(y:pd.Series,
                            alpha:int=1,beta:int=15,
                            lam:Tuple[int,int]=(-5, -2),
                            p:Tuple[int,int]=(-9, -5)
                            ) -> Optional[np.ndarray]:
    """
    Automatically optimizes the baseline of the diffractogram using the cost function defined in the cost_func function.
    It is important to specify the range of hyperparameters lam and p to search for the optimal baseline.
    Parameters:
    - y : pd.Seriesy
        Series containing the diffractogram data.
    - alpha : int, optional
        Weight of the term 1 in the cost function. Default is 2.
    - beta : int, optional
        Weight of the term 2 in the cost function. Default is 10.
    - lam : tuple of int, optional
        Range of lambda values to search for the optimal baseline. Default is (-5, -2).
    - p : tuple of int, optional
        Range of p values to search for the optimal baseline. Default is (-9, -5).

    Returns:
    - np.array or None
        Optimal baseline, or None if the input series is empty or invalid.
    """
    if y.empty:
        logging.warning("Input series is empty.")
        return None


    def filter_monotonic_peaks(y, peaks):
        """Keep only peaks that belong to a monotonically increasing sequence."""
        filtered_peaks = [peaks[0]]
        for i in range(1, len(peaks)):
            if y[peaks[i]] >= y[filtered_peaks[-1]]:
                filtered_peaks.append(peaks[i])
            elif y[peaks[i]] <= y[filtered_peaks[-1]]:
                filtered_peaks[-1]=peaks[i]
                # print('erase last peak')
        return np.array(filtered_peaks)
    # Define a cost function to evaluate the baseline
    def cost_func(y, baseline, alpha=alpha, beta=beta):
        '''
        This is a cost function which tries to fit a basaeline which (Term 1) decreases the distance to the strongest minima and (Term 2) keeps the curvature positive to prevent overfitting the whole diffractogram
        '''
        # Find the strong minima in the data
        inverted_y = -y  # Invert the data to find minima as peaks
        peaks, _ = find_peaks(inverted_y, distance=3)  # Replace 0.5 with an appropriate threshold for your data
        # print(peaks)
        # Filter peaks to get a monotonically increasing sequence
        # filtered_peaks = filter_monotonic_peaks(y, peaks) #!didnt work well
        filtered_peaks=peaks
        # print(filtered_peaks)
        # Term 1: Distance from strong minima (we consider only the found peaks)
        term1 = np.sum(np.abs(y[filtered_peaks] - baseline[filtered_peaks]))

        # # Term 2a: Curvature of the baseline 
        #* didnt work well
        curvature = np.diff(baseline, 2)
        # term2 = np.sum(np.abs(curvature))
        
        # Term 2: Penalty for negative curvature
        term3 = np.sum(np.clip(-curvature, 0, None))

        return alpha * term1  + beta * term3




    # Grid search
    best_cost = float('inf')
    best_lam = None
    best_p = None
    best_baseline = None
    for _lam in np.logspace(lam[0], lam[1], lam[1]-lam[0]+1):
        for _p in np.logspace(p[0], p[1], p[1]-p[0]+1):
            baseline = baseline_als_optimized(y, _lam, _p)
            cost = cost_func(y, baseline)
            if cost < best_cost:
                best_cost = cost
                best_lam = _lam
                best_p = _p
                best_baseline = baseline

    logging.info(f"Best lambda: {best_lam}, Best p: {best_p}") 
    return best_baseline





def structuremodel(pars:Parameters, x:pd.Series, y:pd.Series, n_gaussians:int)->pd.Series:
    """Residual function which is used to fit several gauss functions to the data x,y

    Parameters
    ----------
    - pars : Parameters
        given parameters of the model
    - x : pd.Series
        x values of data
    - y : pd.Series
        y values of data
    - n_gaussians : int
        number of gaussian functions which will be used

    Returns
    -------
    - pd.Series
        returns the residual of fitted model and the actual y values
    """
    model = np.zeros_like(x)
    for i in range(1, n_gaussians + 1):
        amp_key = f'amp{i}'
        cen_key = f'cen{i}'
        sigma_key = f'sigma{i}'
        model += gaussian(x, pars[amp_key], pars[cen_key], pars[sigma_key])
    model += pars['y0']
    return model - y
@log_function_call
def fit_structurefactor(x:pd.Series,y:pd.Series,
                      structure:str,
                      sigmarange:Tuple=(0,0.05),
                      max_sigma_diff:int=None,
                      q_range: Union[str,Tuple[float,float]] ='auto',
                      q111:Union[str,float] ='auto',
                      min_q111:Union[float,None]=None
                      )->MinimizerResult:
    """fits the Structure factor with a model depending on the crystal structure

    Parameters
    ----------
    - x : pd.Series
        q values of the diffractogram
    - y : pd.Series
        structure factor of the diffractogram
    - structure : str
        possibilities: 'fcc'
    - sigmarange : Tuple, optional
        constrain for the min and max possiblee peak width, by default (0,0.05)
    - max_sigma_diff : int, optional
        e.g. 0.1. optional parameter to constrain the maximum relative difference between the peaks, if, by default None
    - q_range : Union[str,Tuple[float,float]], optional
        if 'auto' the function will cut the diffractogram to the first and last tale found, else q range must be provided as tuple, by default 'auto'
    - q111 : Union[str,float], optional
        estimated q value of the (111) peak. if 'auto' is used, the algorithm is using the first peak in the diffractorgram, by default 'auto'
    Returns
    -------
    - Tuple[pd.Series,MinimizerResult]
        overall fit and MinimizerResult class containing the fit parameters, such as peak widths and positions
    """
    if q111=='auto' :
        peaks=find_peaks(y, prominence=0.1)
        if len(peaks)==0:
            logging.info(f'found {len(peaks)} peaks in y {y}')
        peaks=peaks[0]
        if min_q111 is not None:
            peaks=[p for p in peaks if x.iloc[p]>min_q111]
        logging.info(f'maxima found at positions: {x.iloc[peaks]}')
        q111=x.iloc[peaks[0]]*0.99 #estimate for the first position of q
        
        logging.info(f'guess of the (111) peak position is q = {q111}')
    
    #determine max q depending on the amount of peaks to fit, adjust 
    peakpositions=indcate_diffractionpattern(q111,structure=structure,ax=None)
    logging.info(f'assumed peaks at positions: {peakpositions}')
    if q_range == 'auto':
        x,y = auto_cutdiffr(x,y,peakpositions[-1])
        logging.info(f'automatically cut the diffractorgram to the first and last tale found')
        
    elif type(q_range)==tuple:
        df=pd.DataFrame([x,y])
        df=cut_diffractogram(df,*q_range)
        x,y=df.iloc[:,0],df.iloc[:,1]

    else:
        logging.error(f'q_range must be tuple of floats or "auto", not {type(q_range)}')
    #initiate the restructions acc. to crystal structure
    pfit = Parameters()

    if structure=='fcc':
         ####number of gauss functions
        #in fcc the relative peak positions for the peak (j k l) is e.g. q111*(j + k + l)*(1/3)**(0.5)
        relativepositions=[3,4,8,11,12,16,19,20,24,27] #TODO adjust this later to just provide a list of possible signals depending of the structure
        n_gaussians=len(relativepositions)
        expression= '/3)**(0.5)'   
        mathexpression= lambda i,q: q*(i/3)**(0.5)   

        
        
    else:
        logging.info('unknown crystal structure given as input')
        return None

    
    for i in range(n_gaussians):
        if i==0:
            pfit.add(name='cen'+str(i+1), value=q111,min=q111*0.93,vary=True)
            pfit.add(name=f'sigma{i+1}', value=0.03, min=0, max=0.05)
        else:

            q=mathexpression(relativepositions[i],q111)
            logging.info(f'try adding gaussian no {i+1} at q = {q}')
            if q>x.iloc[-1]: 
                logging.info(f'q={q} is larger than the maximum q={x.iloc[-1]} - breaking')
                imax=i
                break # breaks the loop in order to not put more peaks as theoretically required
            pfit.add(name='cen'+str(i+1), expr='cen1*('+str(relativepositions[i])+expression)
            # Set initial value and constraints for sigma
            if max_sigma_diff is not None:
                # Add constraints to ensure sigma values are within max_sigma_diff of each other
                pfit.add(f'sigma_diff{i+1}', value=0, min=-max_sigma_diff, max=max_sigma_diff) 
                pfit.add(name=f'sigma{i+1}', expr=f'sigma1*(1+sigma_diff{i+1})')
    
            else:    
                pfit.add(name=f'sigma{i+1}', value=(sigmarange[1]-sigmarange[0])/2+sigmarange[0], min=sigmarange[0], max=sigmarange[1]) 
            
           
        pfit.add(name='amp'+str(i+1), value=0.03,min=0,max=0.1)
        # pfit.add(name='sigma'+str(i+1), value=0.03,min=0,max=0.1)

    pfit.add(name='y0',min=0, value=0,max=0.02)
    if not imax: imax=n_gaussians
    logging.info(f'number of gaussians={imax}')
    mini = Minimizer(structuremodel, pfit, fcn_args=(x, y, imax))
    res = mini.leastsq()
    res.rsquared=1 - res.residual.var() / np.var(y)  # needs to be added, as lmfit  has it not included. -> can be called using out.rsquared.
    logging.info(f'fit successfull with R2 = {res.rsquared:2f}')
    # best_fit = y + out.residual #optional but can also be calculated later if required
    return res 

@log_function_call
def indcate_diffractionpattern(q111:float,
                      structure:str,
                      ax=None,
                      )->List[float]:
    
    

    if structure=='fcc':

        #in fcc the relative peak positions for the peak (j k l) is e.g. q111*(j + k + l)*(1/3)**(0.5)
        relativepositions=[3,4,8,11,12,16,19,20,24,27] #TODO adjust this later to just provide a list of possible signals depending of the structure

        expression= lambda i,q: q*(i/3)**(0.5)   
        
    else:
        logging.info('unknown crystal structure given as input')
        return None
    

    peakpositions=[]
    for i in relativepositions:
        peakpositions.append(expression(i,q111))
        if ax is not None: ax.axvline(expression(i,q111),color='red',linewidth=1,alpha=0.5)
    return peakpositions

def get_structureinfo(out:MinimizerResult,
                      structure:str='fcc'
                      )->Tuple[float,float]:
    """calculates the lattice constant and domain size from the fitted parameters.
    The function returns a tuple containing the lattice constant and domain size, respectively.
    
    Parameters
    ----------
    -  out : MinimizerResult
        Fit result obtained using the fit_structurefactor() function.
    
    - structure : str, optional
        crystal structure, by default 'fcc'

    Returns
    -------
    - Tuple[float,float]
        lattice constant and domain size in nm
    """
    if structure=='fcc':
        a=2*3**0.5*np.pi/out.params["cen1"].value
        fwhm=2*out.params["sigma1"].value*np.sqrt(2*np.log(2))
        Dsize=2*np.pi/fwhm
    else:
        raise('unknown crystal structure given as input')
    return a, Dsize








# Define form factor functions
def spherical_form_factor(q, R):
    qr = q * R
    return (3 * (np.sin(qr) - qr * np.cos(qr)) / qr**3) ** 2

def cubic_form_factor(q, R):
    return np.sinc(q * R)**2

# Main scattering intensity function

def formfactor(q, mu, sigma, C, y0,distribution, shape,distribution_type='volume' ):
    amp=1
    if distribution == 'lognormal':
        mu = np.log(mu)
        size_distribution = lognormal
    elif distribution == 'normal':
        sigma = sigma * mu
        size_distribution = gaussian

    if shape == 'spherical':
        form_factor = spherical_form_factor
    elif shape == 'cubic':
        form_factor = cubic_form_factor

    if distribution_type == 'volume':
        intensity_func = lambda R: 10**C * form_factor(q, R) * size_distribution(R,amp, mu, sigma) + y0
    elif distribution_type == 'number':
        intensity_func = lambda R: 10**C * form_factor(q, R) * size_distribution(R,amp, mu, sigma) * R**3  + y0
    else:
        raise ValueError("Invalid distribution type. Choose 'volume' or 'number'.")
    upper_limit = 10 * mu  # Example: 10 times the mean size
    lower_limit = 1e-3    # A small positive number to avoid division by zero issues
    return integrate.quad(intensity_func,  lower_limit, upper_limit)[0]

formfactor=np.vectorize(formfactor)
@log_function_call                                  
def fit_formfactor(q_data, I_data,initial_params, distribution='lognormal', shape='spherical', distribution_type='volume'):

    def model(q, mu, sigma, C, y0):
        # mu, sigma, C, y0=initial_params
        
        return formfactor(q,mu, sigma, C, y0,distribution, shape,distribution_type)
    logging.info(f'fitting formfactor model to SAXS data...')
    #* fit the log is required to fit optimally over several orders of magnitude
    popt, pcov = curve_fit(lambda q,mu, sigma, C, y0:np.log(model(q,mu, sigma, C, y0)), q_data, np.log(I_data), p0=initial_params)
    # calculate the R²
    residuals=np.log(I_data) - np.log(model(q_data, *popt))
    ss_res = np.sum(residuals**2)
    ss_tot = np.sum((np.log(I_data)-np.mean(np.log(I_data)))**2)
    r_squared = 1 - (ss_res / ss_tot)
    logging.info(f'fit successfull with R²={r_squared:.3f}')
    if distribution=='normal':fitresult=f'mean size: {popt[0]:.2f} nm +/- {popt[1]*100:.1f} %'
    elif distribution=='lognormal':fitresult=f'mean size: {np.exp(np.log(popt[0])+popt[1]**2/2):.2f} nm, mode size: {np.exp(np.log(popt[0])-popt[1]**2):.2f} nm +/- {popt[1]*100:.1f} %'
    logging.info(fitresult)
    return popt, pcov,r_squared, fitresult
@log_function_call 
def plot_formmodel(q_data,  params, distribution='lognormal', shape='spherical', distribution_type='volume', ax=None, color='red',label=None):
    def model(q, mu, sigma, C, y0):
        # mu, sigma, C, y0=initial_params
        return formfactor(q,mu, sigma, C, y0,distribution, shape,distribution_type)
    if ax is None:
        fig, ax = plt.subplots()
        ax.set_xlabel(r'Scattering vector, q ($\mathregular{nm^{-1}}$)')
        ax.set_ylabel('Intensity (a.u.)')
        ax.set_yscale('log')
    # ax.plot(q_data, I_data, 'o', label='data')
    p=ax.plot(q_data, model(q_data,*params),color=color, label=label)
    # ax.set_xlabel('Scattering vector, q ($\mathregular{nm^{-1}}$)')
    # ax.set_ylabel('Intensity (a.u.)')
    # ax.set_yscale('log')
    if label: ax.legend()
    return p


#* Data visualization

@log_function_call
def plot(df: pd.DataFrame, x: str = 'q', y: str = 'I', 
         subtractbaseline=False,
         ax: Optional[plt.Axes] = None, scale: str = 'log',
         line:bool=True,
         label: Optional[str] = None)->Tuple[matplotlib.lines.Line2D,Optional[matplotlib.lines.Line2D]]:
    """
    Plots the data based on the specified columns and settings.

    Parameters:
    - df : pd.DataFrame
        DataFrame containing the data to plot.
    - x : str, optional
        Column name for the x-axis. Default is 'q'.
    - y : str, optional
        Column name for the y-axis. Default is 'I'.
    - subtractbaseline : bool, optional
        Whether to subtract the baseline from the data. Only works when y='S' and a baseline is already created. Default is False.
        If a baseline is in the DataFrame, the baseline will be plotted when y='S'.
    - ax : matplotlib.pyplot.Axes, optional
        Axis on which to plot the data. If None, a new subplot is created.
    - scale : str, optional
        Scale for the y-axis. Default is 'log'.
    - line : bool, optional
        if False, function returns a scatterplot, else a 2DLine plot
    - label : str, optional
        Label for the curve on the plot.

    Returns:
    - Tuple[matplotlib.lines.Line2D, Optional[matplotlib.lines.Line2D, None]]
        Tuple containing the line object and the baseline line object.
        The baseline line object is None if the 'baseline' column is not present in the DataFrame.
    """

    # Check if essential columns exist in DataFrame
    if x not in df.columns or y not in df.columns:
        print(f"Error: DataFrame does not contain necessary columns {x} and/or {y}.")
        return

    # Create a new subplot if no axis is specified
    if ax is None:
        fig, ax = plt.subplots()
    
    if (subtractbaseline==True) & (y=='S'):
        if ('baseline' in df.columns):
            y_= df[y] - df['baseline']
        else: logging.warning(f"No baseline column found in DataFrame.")
    else: y_=df[y]
    
    # Plot the data
    if line==True:
        p=ax.plot(df[x], y_, label=label)
        p=p[0]
    else: p=ax.scatter(df[x], y_, label=label)
    if (y=='S') & ('baseline' in df.columns)&(subtractbaseline==False):
        b=ax.plot(df[x], df['baseline'],color='red', label='baseline')
    else: b=None
    # Set the y-axis scale
    ax.set_yscale(scale)

    # Set axis labels
    y_label_map = {'I': 'Intensity (a.u.)','F': 'Intensity (a.u.)', 'S': 'Structure Factor (a.u.)'}
    y_label = y_label_map.get(y, y)
    ax.set_xlabel(r'Scattering vector ($\mathregular{nm^{-1}}$)')
    ax.set_ylabel(y_label)

    # Add legend if label is specified
    if label:ax.legend()

    
    if b is not None:
        return p,b
    else:
        return p
     
@log_function_call
def plot_structuremodel(x_array:np.array,out:MinimizerResult, 
             peakcolor:str='green',fitcolor:str='black',
             ax:Optional[plt.Axes]=None,
             label: Optional[str] = None,
             yshift:int=0
             )->List[matplotlib.lines.Line2D]:
    """plots the fitted model using the parameters from the MinimizerResult class.
    The function returns a list of the plots, which can be used to further customize the plot.
    
    Parameters
    ----------
    - x_array : np.array
        data array for the x axis
    - out : MinimizerResult
        Fit results provided by the function fit_structurefactor()
    - peakcolor : str, optional
        color of the single gauss functions, by default 'green'
    - fitcolor : str, optional
        color of the overall fit, by default 'black'
    - ax : plt.Axes, optional
        in case of given axis shall be used, otherwise a new one is created., by default None
    - label: str, optional
        label for the overall fit, by default None
    Returns
    -------
    - List
        List of matplotlib.lines.Line2D and matplotlib.collections.PolyCollection containing the plots of gauss functions and area fillings, respectively. Allows for subsequent adjustments.
    """
    if ax is None: fig, ax =plt.subplots()

    y0=out.params['y0'].value
    # n_gaussians=int(list(out.params)[-2].replace('sigma',''))
    # logging.info(f'number of gaussians is {n_gaussians}')
    plots=[]
    gauss_peak_sum=np.array([y0]*len(x_array))

    logging.info(f'plotting gauss functions...')
    for i in range(len(out.params)):
        try:
            amp=out.params['amp'+str(i+1)].value
            cen=out.params['cen'+str(i+1)].value
            sigma=out.params['sigma'+str(i+1)].value
            gauss_peak = gaussian(x_array, amp, cen, sigma)
            gauss_peak_sum=gauss_peak_sum+gauss_peak
            peak_=ax.plot(x_array, gauss_peak+yshift, peakcolor)
            area_=ax.fill_between(x_array, gauss_peak.min()+yshift, gauss_peak+yshift, facecolor=peakcolor, alpha=0.1)
            plots.append((peak_,area_))
        except Exception as e:
            logging.info(f'stopped plotting at {e}')
            break
    l_=ax.plot(x_array, gauss_peak_sum+yshift, color=fitcolor,label=label)
    plots.append(l_)
    if label:ax.legend()
    return plots



# %%
if __name__=='__main__':
    pass
    


